src/common/mutex_debug.cc

```diff
@@ -25,15 +25,12 @@ enum {
   l_mutex_last
 };
 
-mutex_debugging_base::mutex_debugging_base(const std::string &n, bool bt)
-  : name(n), id(-1), backtrace(bt), nlock(0), locked_by(thread::id())
-{
-  if (g_lockdep)
-    _register();
-}
-mutex_debugging_base::mutex_debugging_base(const char *n, bool bt)
-  : name(n), id(-1), backtrace(bt), nlock(0), locked_by(thread::id())
+mutex_debugging_base::mutex_debugging_base(std::string n, bool bt)
+  : name(std::move(n)), id(-1), backtrace(bt), nlock(0), locked_by(thread::id())
 {
+  if (name.empty())
+    name = std::to_string(reinterpret_cast<const uintptr_t>(this));
```

Comment:
I think we ultimately should alter `lockdep` to register mutexes basing on an unique trait (like instance's address) instead of name. In many situations names are not unique. We actually have e.g. constructor for `shared_mutex_debug` that explicitly uses empty string for it.

Reply:
If you use instances instead of names then you lose most of the power of lockdep, since it won't be as effective in identifying possible lock cycles.

Reply:
Instead, we should fix the callers that pass an empty string!

Reply:
Per our talk on the stand-up: currently using the `name` causes `lockdep` to track dependencies between multiple and **different** instances of e.g. `pthread_mutex_t`. In different words: `lockdep` groups single mutexes into a class identified by `name`. This is the intended behavior. When a false-positive can happen because of that, a caller is obliged to guarantee name uniqueness on its own.